src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+                            {
+                                typeof (ArgumentException),
+                                typeof (NotSupportedException)
+                            };
+                        else
+                            expectedExceptions = new[]
+                            {
+                                typeof (ArgumentException)
+                            };
+
+                    object invalid1 = invalid;
+                    AssertThrows(expectedExceptions, () => list[0] = invalid1);
+                    CollectionAssert.Equal(items, list);
+                }
+            }
+        }
+
+        [Fact]
+        public void SetItem()
+        {
+            if (!ExpectedIsReadOnly)
+            {
+                if (ItemsMustBeNonNull)
+                {
+                    var items = GenerateItems(32);
+                    var list = GetList(items);
+                    Assert.Throws<ArgumentNullException>(
+                        () => list[0] = null);
+                    CollectionAssert.Equal(items, list);
+                }
+                else
+                {
+                    var items = GenerateItems(32);
+                    var list = GetList(items);
+                    list[0] = null;
+                    items[0] = null;
+                    CollectionAssert.Equal(items, list);
+
+                    items = GenerateItems(32);
+                    list = GetList(items);
+                    list[list.Count - 1] = null;
+                    items[items.Length - 1] = null;
+                    CollectionAssert.Equal(items, list);
+                }
+
+                if (ItemsMustBeUnique)
+                {
+                    var items = GenerateItems(32);
+                    var list = GetList(items);
+                    Assert.Throws<ArgumentException>(
+                        () => list[0] = items[1]);
+                    CollectionAssert.Equal(items, list);
+                }
+                else
+                {
+                    var items = GenerateItems(32);
+                    var list = GetList(items);
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        list[i] = items[items.Length - i - 1];
+                    }
+                    Array.Reverse(items);
+                    CollectionAssert.Equal(items, list);
+                }
+
+                {
+                    var items = GenerateItems(32);
+                    var origItems = (object[]) items.Clone();
+                    var list = GetList(items);
+
+                    // verify setting every item
+                    for (int i = 0; i < items.Length; i++)
+                    {
+                        if (i < items.Length/2)
+                        {
+                            list[items.Length - i - 1] = GenerateItem();
+                        }
+                        list[i] = items[items.Length - i - 1];
```

Comment:
I think this could use a comment saying why it doesn't need to check ItemsMustBeUnique.  "list" and "items" are so similar of words that I thought it was replacing the back half of the list, then copying it to the front half.